src/browser.cc

```diff
@@ -0,0 +1,182 @@
+       CEREAL_NVP(total),
+       CEREAL_NVP(browse_uri),
+       CEREAL_NVP(browse_all_uri));
+  }
+};
+
+struct InsightResponse {
+  std::string prev_uri;
+  std::string next_uri;
+  std::set<std::string> tags;
+  double score;
+  std::string description;
+  std::unique_ptr<insight::AbstractBase> insight;
+  InsightResponse(const InsightsOutput& input, size_t index) {
+    if (index) {
+      prev_uri = FLAGS_output_uri_prefix + "/?id=" + ToString(index);  // It's 1-based in the URI.
+    }
+    if (index + 1 < input.insight.size()) {
+      next_uri = FLAGS_output_uri_prefix + "/?id=" + ToString(index + 2);  // It's 1-based in the URI.
+    }
+    score = input.insight[index]->score;
+    input.insight[index]->EnumerateFeatures([this, &input](const std::string& feature) {
+      const auto cit = input.feature.find(feature);
+      assert(cit != input.feature.end());
+      tags.insert(cit->second.tag);
+      assert(input.tag.find(cit->second.tag) != input.tag.end());
+    });
+    description = input.insight[index]->Description();
+    insight = CloneSerializable(input.insight[index]);
+  }
+  template <typename A>
+  void serialize(A& ar) {
+    ar(CEREAL_NVP(prev_uri),
+       CEREAL_NVP(next_uri),
+       CEREAL_NVP(tags),
+       CEREAL_NVP(score),
+       CEREAL_NVP(description),
+       CEREAL_NVP(insight));
+  }
+};
+
+struct SmartSessionInfo {};
+
+int main(int argc, char** argv) {
+  ParseDFlags(&argc, &argv);
+
+  if (FLAGS_route.empty() || FLAGS_route.back() != '/') {
+    std::cerr << "`--route` should end with a slash." << std::endl;
+    return -1;
+  }
+
+  const auto input = ParseJSON<InsightsOutput>(FileSystem::ReadFileAsString(FLAGS_input));
+
+  HTTP(FLAGS_port).Register(FLAGS_route, [&input](Request r) {
+    const auto one_based_index = FromString<size_t>(r.url.query["id"]);
+    if (one_based_index && one_based_index <= input.insight.size()) {
+      if (!r.url.query["html"].empty()) {
+        using namespace html;
+        HTML html_scope;
+        {  // HEAD.
+          HEAD head;
+          TITLE("Insights Visualization Alpha");
+        }
+        {  // TABLE, TR.
+          TABLE table({{"border", "0"}, {"align", "center"}, {"cellpadding", "8"}});
+          TR r({{"align", "center"}});
+          if (one_based_index > 1) {
+            TD d;
+            A a({{"href", FLAGS_route + "?id=" + ToString(one_based_index - 1) + "&html=yes"}});
+            TEXT("Previous insight");
+          }
+          if (one_based_index < input.insight.size()) {
+            TD d;
+            A a({{"href", FLAGS_route + "?id=" + ToString(one_based_index + 1) + "&html=yes"}});
+            TEXT("Next insight");
+          }
+        }  // TABLE, TR.
+        input.insight[one_based_index - 1]->RenderHTML(input.feature);
+        r(html_scope.AsString(), HTTPResponseCode.OK, "text/html");
+      } else {
+        r(InsightResponse(input, one_based_index - 1));
+      }
+    } else if (r.url.query["id"] == "all") {
```

Comment:
Not sure I understand. It'll be zero, which is not within `[1 .. N]` -- appears logically innocent to me.